lib/improver/utilities/indexing_operations.py

```diff
@@ -0,0 +1,168 @@
+
+       The Array column gives the index in the outer most dimension of the
+       multi-dimensional input array (array_set). The Data column shows the
+       contents of the 3 sub-arrays that correspond to the given Array index.
+       The Indices examples are used to show which data would be returned
+       (Result) should the index_array be shaped in these various ways, with
+       these array index values. The colour coding is used to help make clear
+       how the shape of the index_array relates to the arrays within the
+       array_set.
+
+    The examples above demonstrate the application of choose with a series of
+    index_arrays of different dimensionality. In all but the last of these
+    cases the choose function will use broadcasting to construct an index array
+    that matches the shape of the sub-arrays from which data is being taken.
+    The figure below illustrates the results of this intermediate broadcasting
+    step.
+
+    .. figure:: extended_documentation/utilities/indexing_operations/
+       numpy_choose_broadcasting.png
+       :scale: 60 %
+       :alt: Diagram to help explain broadcasting behaviour of choose.
+
+       The form of the index_arrays after they have been broadcast to match the
+       sub-arrays from which values are being taken.
+
+    Args:
+        index_array (np.array of ints):
+            This array must contain integers in the range [0, n-1], where n is
+            the number of arrays in the array_set (equivalent to the length of
+            the leading dimension of array_set if it is a multi-dimensional
+            array rather than a list).
+        array_set (np.array):
+            A multi-dimensional array, where the leading dimension is in effect
+            an indexing dimension. Within this leading dimension are the
+            sub-arrays from which values are to be extracted at positions that
+            match those given in the index_array.
+    Returns:
+        result (np.array):
+            An array containing data extracted from the array_set. The returned
+            array will have the same shape as the sub-arrays unless the
+            index_array has dimensions equivalent to array_set. In this case
+            the returned array will have an equivalent number of dimensions
+            (i.e. one more than the sub-arrays).
+    Raises:
+        ValueError: If input arrays cannot be broadcast to a common shape.
+        IndexError: If dimensions are added to array_set during the
+                    broadcasting of arrays; this will prevent the later
+                    indexing from working as expected.
+        IndexError: If an index exceeds the number of available sub-arrays.
+    """
+    # Used to correct array shape in output following intermediate broadcast.
+    index_array = np.array(index_array)
+    index_dims = index_array.ndim
+    trim_result = index_array.shape != array_set.shape
```

Comment:
When would this happen?  Is `!=` the correct condition here (ie is the expected case always less or always more, or are there several possible acceptable cases)?

Reply:
Removed.